### PR TITLE
HIGGS Quantization Support

### PR DESCRIPTION
**Pull request by BlackSamorez**
### HIGGS 0-Shot Quantization

HIGGS is a new 0-shot quantization algorithm that combines Hadamard preprocessing with MSE-Optimal quantization grids to achieve lower quantization error and SOTA performance. You can find more information in the [paper](https://arxiv.org/abs/2411.17525).

Runtime support for HIGGS is implemented through [FLUTE](https://arxiv.org/abs/2407.10960), and its [library](https://github.com/HanGuo97/flute?tab=readme-ov-file).

This PR **adds support for HIGGS+FLUTE into `transformers` allowing for low-error 0-shot quantization and fast LLM inference**.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @SunMarc @MekkCyber

**Comment by SunMarc** [comment]
cc @MekkCyber

**Comment by BlackSamorez** [comment]
Failed tests look like a problem on the runner's end

File: src/transformers/utils/import_utils.py
Comment (SunMarc) [inline src/transformers/utils/import_utils.py:110]: let's do the version check in `is_flute_available` and you shouldn't pin the version like that. We should allow a minimum version.

File: src/transformers/utils/quantization_config.py
Comment (SunMarc) [inline src/transformers/utils/quantization_config.py:1247]: Can you use instead `modules_to_not_convert` just like the other quantization method ?

File: src/transformers/utils/quantization_config.py
Comment (SunMarc) [inline src/transformers/utils/quantization_config.py:1243]: What are the accepted values ?

File: src/transformers/utils/quantization_config.py
Comment (SunMarc) [inline src/transformers/utils/quantization_config.py:1245]: Could you explain a bit more what this is used for ?

File: src/transformers/utils/quantization_config.py
Comment (SunMarc) [inline src/transformers/utils/quantization_config.py:1263]: Can you add a description of what this is used for ? The user shouldn't have to worry about that ?

File: src/transformers/utils/quantization_config.py
Comment (SunMarc) [inline src/transformers/utils/quantization_config.py:1272]: add in post_init checks for bits and p

File: tests/quantization/higgs/test_higgs.py
Comment (SunMarc) [inline tests/quantization/higgs/test_higgs.py:60]: to remove or uncomment

File: tests/quantization/higgs/test_higgs.py
Comment (SunMarc) [inline tests/quantization/higgs/test_higgs.py:68]: can we use a smaller model like a tiny llama ? This will be better for our CI thanks !

File: tests/quantization/higgs/test_higgs.py
Comment (SunMarc) [inline tests/quantization/higgs/test_higgs.py:113]: to remove or perform some tests with this device_map. I think we shouldn't allow users to pass this kind of device_map and some check should be added in `validate_environment`. Check for example the awq integration code

**Review by SunMarc — COMMENTED** [review]
Thanks for integrating this new quantization method so fast! I left some comments and don't forget to also update the documentation so that the users knows how to use it !

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:158]: Added comments to this and possible repacking happening afterwards

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:51]: Just a small nit to help users
```suggestion
        raise NotImplementedError(f"Device capability {target_device_cc} not supported for FLUTE (yet?) to verify your device capability check out https://developer.nvidia.com/cuda-gpus")
```

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:129]: Just for my understanding why do we need the `num_sms_packed` ?

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:173]: Just a small question, is the `group_size` a constant ?

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:219]: Do you mean that serialization is not implemented yet ? so we can't save a quantized model and load it ?

File: tests/quantization/higgs/test_higgs.py
Comment (MekkCyber) [inline tests/quantization/higgs/test_higgs.py:155]: I think you meant `nb_higgs_linear` 😉

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:40]: sorry if i'm mistaken, I don't believe we use this function anywhere

**Comment by MekkCyber** [comment]
Hey @BlackSamorez, Thanks for adding this quantization method so quickly ! I added some very small nits

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:129]: Codes packing is `sms` dependent. We need to remember what was the `sms` of the machine on which the codes were packed on (`num_sms_packed`) to be able to check if we need to repack or not. Moreover, we need `num_sms_packed` to do the repacking itself.

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:219]: No, serialization is fully functional. This message got copied with bnb code I borrowed and I forgot to remove it.
By the way, bnb implemented serialization quite some time ago as well.

File: src/transformers/integrations/higgs.py
Comment (BlackSamorez) [reply src/transformers/integrations/higgs.py]: Removed

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py]: Done

File: src/transformers/utils/quantization_config.py
Comment (BlackSamorez) [reply src/transformers/utils/quantization_config.py]: Updated the docstring to better reflect what those are

File: tests/quantization/higgs/test_higgs.py
Comment (BlackSamorez) [reply tests/quantization/higgs/test_higgs.py:68]: Sadly, no. FLUTE is only compiled for specific matrix shapes, for now.
TinyLlama is not among those shape. Nor is any model smaller than 8B.

File: tests/quantization/higgs/test_higgs.py
Comment (BlackSamorez) [reply tests/quantization/higgs/test_higgs.py]: Sure. Fixed

File: tests/quantization/higgs/test_higgs.py
Comment (BlackSamorez) [reply tests/quantization/higgs/test_higgs.py]: Uncommented this

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:50]: I added `and in-flight quantization of full-precision models`.

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py]: Removed the unused function. Thanks!

File: src/transformers/utils/quantization_config.py
Comment (BlackSamorez) [reply src/transformers/utils/quantization_config.py]: Added explanation

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (BlackSamorez) [reply docker/transformers-quantization-latest-gpu/Dockerfile]: Thanks, updated.

File: src/transformers/utils/quantization_config.py
Comment (BlackSamorez) [reply src/transformers/utils/quantization_config.py]: Added more info

File: src/transformers/utils/import_utils.py
Comment (BlackSamorez) [reply src/transformers/utils/import_utils.py]: Done

**Comment by BlackSamorez** [comment]
@SunMarc @MekkCyber thanks for your feedback!
I think I addressed all of your concerns.

**Comment by MekkCyber** [comment]
Hey @BlackSamorez, thanks for iterating and integrating this method 🤗 ! LGTM ! Can you add a documentation in `docs/source/en/quantization/higgs.md` to explain a bit how Higgs works, why it only works for models > 8B for now, and how to load a quantized model ? Also i left a small question

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:219]: Right, that's a legacy comment from bnb code. Forgot to change it. Done. Thanks!

**Comment by BlackSamorez** [comment]
Added documentation

File: docs/source/en/quantization/higgs.md
Comment (MekkCyber) [inline docs/source/en/quantization/higgs.md:16]: Thanks for the docs 🔥 ! Let's also add a simple example of how to quantize models using Higgs even if it looks very simple

**Comment by BlackSamorez** [comment]
Added usage example in docs.
Also, I had to slightly rework the`HiggsConfig` to add more quantization parameters. It turned out that the previous hard- coded setup didn't work for `gemma` models. Now it should be fine.

File: src/transformers/quantizers/quantizer_higgs.py
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_higgs.py:90]: Sorry I didn't see this earlier 🙏 , here we need a `logger.info` like in other quantizers to inform the user which dtype we are setting and why

**Review by MekkCyber — APPROVED** [review]

**Comment by BlackSamorez** [comment]
> Thank you @BlackSamorez ! Looks great to me ! Just out of curiosity what was the reason why hard-coding the `group_size` don't work with `gemma`, it looks the same

It wasn't `groups_size`, it was `hadamard_size`. We have to pad matrices to a multiple of it. It was 1024, but gemma internal dimensions are not divisible by 1024. Padding occurred, and pre-compiled kernels stopped working for new shapes.
Lowering `hadamard_size` to 512 by default fixed the issue.

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py]: Thanks for noticing. I added more logging and asserts, as well as fixed bf16 support elsewhere

**Comment by BlackSamorez** [comment]
@SunMarc Hi! Is there a chance you could take a look at it this week? We wanted to release it before NIPS.

**Comment by SunMarc** [comment]
> @SunMarc Hi! Is there a chance you could take a look at it this week? We wanted to release it before NIPS.

I'll do that now. When is the deadline ? After my review, I still need this to be reviewed by a core maintainer

**Comment by BlackSamorez** [comment]
> When is the deadline ?

There isn't a hard deadline, but it would be very nice to have this merged this week or early next week.

File: tests/quantization/higgs/test_higgs.py
Comment (SunMarc) [reply tests/quantization/higgs/test_higgs.py:68]: sounds good ! Do we have a check somewhere for the matrix shape then ? If a run this on a tinyllama, what kind of error will I get. If it is possible to detect the issue beforehand and output a nice error message, that would be nice.

**Review by SunMarc — APPROVED** [review]
Thanks for your work ! Just a few nits.

File: docs/source/en/quantization/higgs.md
Comment (BlackSamorez) [reply docs/source/en/quantization/higgs.md]: I meant PEFT. I'm familiar with the process of integrating quantizations into the PEFT library: I did it for AQLM.

File: src/transformers/quantizers/quantizer_higgs.py
Comment (BlackSamorez) [reply src/transformers/quantizers/quantizer_higgs.py:168]: It's a bunch of transposes and memory layout changes. It should be fast.

File: docs/source/en/quantization/higgs.md
Comment (BlackSamorez) [reply docs/source/en/quantization/higgs.md:41]: Sadly, T4 is not among the supported GPUs yet, so no Colab demo for now.

**Comment by BlackSamorez** [comment]
Pre-quantized a bunch of models (including `Llama-3.3-70B-Instruct`) and added a link to the collection to docs.

File: docs/source/en/quantization/higgs.md
Comment (SunMarc) [reply docs/source/en/quantization/higgs.md]: Make sense thanks ! It would be nice to add a section to precise which specific gpu works with it

**Comment by SunMarc** [comment]
Gentle ping @ArthurZucker

**Comment by SunMarc** [comment]
Can you fix the CI with `make style` for the quality and potentially rebase the PR ?

**Review by SunMarc — APPROVED** [review]
Thanks for adding new features !

File: docs/source/en/quantization/higgs.md
Comment (ArthurZucker) [inline docs/source/en/quantization/higgs.md:55]: nice!

File: src/transformers/utils/quantization_config.py
Comment (ArthurZucker) [inline src/transformers/utils/quantization_config.py:1281]: we usually go with raising error, let's just switch that!

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot for your hard work and for introducing this new method! 
Super small nit and @SunMarc will merge 🤗

File: docs/source/en/quantization/higgs.md
Comment (BlackSamorez) [reply docs/source/en/quantization/higgs.md:55]: Actually, that was just fixed with `flute-kernel==0.3.0`. Let me update that...

**Comment by BlackSamorez** [comment]
Things to recheck later: 
 - `make style` doesn't actually fix `python utils/custom_init_isort.py --check_only` errors. I had to run `python utils/custom_init_isort.py` manually.
 - `pillow` (`PIL`) is needed for `make quality` but isn't included in `pip install -e .[quality]`

File: src/transformers/utils/quantization_config.py
Comment (BlackSamorez) [reply src/transformers/utils/quantization_config.py]: Done

**Review by SunMarc — APPROVED** [review]
Thanks for iterating ! I will merge this when the CI is green

**Comment by BlackSamorez** [comment]
Quantization tests are nightly, right?

**Comment by SunMarc** [comment]
That's right !

**Pull request merged**
